src/mame/marion/ms68k.cpp

```diff
@@ -0,0 +1,319 @@
+		, m_maincpu(*this, "maincpu")
+		, m_duart(*this, "duart")
+		, m_terminal(*this, "terminal")
+		, m_modem(*this, "modem")
+		, m_fdc(*this, "fdc")
+		, m_floppy(*this, "fdc:%u", 0)
+		, m_floppy_selected_drive(0)
+		, m_scsi(*this, "scsibus:7:ncr5380")
+		, m_scsibus(*this, "scsibus")
+		, m_printer_conn(*this, "prn")
+		, m_printer_out(*this, "prn_out")
+	{ }
+
+	void ms68k(machine_config &config) ATTR_COLD;
+
+	DECLARE_INPUT_CHANGED_MEMBER(nmi_button);
+
+protected:
+	virtual void machine_reset() override ATTR_COLD;
+
+private:
+	void mem_map(address_map &map) ATTR_COLD;
+
+	memory_view m_bootvect;
+
+	required_device<cpu_device> m_maincpu;
+
+	required_device<mc68681_device> m_duart;
+	required_device<rs232_port_device> m_terminal;
+	required_device<rs232_port_device> m_modem;
+
+	required_device<wd1772_device> m_fdc;
+	required_device_array<floppy_connector, 2> m_floppy;
+	int m_floppy_selected_drive;
+
+	required_device<ncr5380_device> m_scsi;
+	required_device<nscsi_bus_device> m_scsibus;
+
+	required_device<centronics_device> m_printer_conn;
+	required_device<output_latch_device> m_printer_out;
+
+	uint8_t duart_r(offs_t offset);
+	void duart_w(offs_t offset, uint8_t data);
+
+	// Floppy
+	static void floppy_types(device_slot_interface &device);
+	static void floppy_formats(format_registration &fr);
+	void floppy_side_w(int state);
+	void floppy_drive0_w(int state);
+	void floppy_drive1_w(int state);
+	void floppy_drive_select(int drive);
+
+	// Printer
+	uint8_t printer_r(offs_t offset);
+	void printer_w(offs_t offset, uint8_t data);
+};
+
+/* Input ports */
+static INPUT_PORTS_START( ms68k )
+	PORT_START("ABORT")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_BUTTON1) PORT_NAME("NMI button") PORT_CHANGED_MEMBER(DEVICE_SELF, ms68k_state, nmi_button, 0)
+INPUT_PORTS_END
+
+
+void ms68k_state::ms68k(machine_config &config)
+{
+	M68000(config, m_maincpu, 16_MHz_XTAL / 2);
+	m_maincpu->set_addrmap(AS_PROGRAM, &ms68k_state::mem_map);
+
+	// Set up DUART.
+
+	MC68681(config, m_duart, 16_MHz_XTAL / 2);
+	// DUART Interrupt = 3
+	m_duart->irq_cb().set_inputline(m_maincpu, M68K_IRQ_3);
+
+	// Set up terminal RS-232.
+
+	RS232_PORT(config, m_terminal, default_rs232_devices, "terminal");
+	m_duart->a_tx_cb().set(m_terminal, FUNC(rs232_port_device::write_txd));
+	m_terminal->rxd_handler().set(m_duart, FUNC(mc68681_device::rx_a_w));
+	// DUART IP0 = CTS
+	m_terminal->cts_handler().set(m_duart, FUNC(mc68681_device::ip0_w));
+	// DUART OP0 = RTS
+	m_duart->outport_cb().append(m_terminal, FUNC(rs232_port_device::write_rts)).bit(0);
+
+	// Set up modem RS-232.
+
+	RS232_PORT(config, m_modem, default_rs232_devices, nullptr);
+	m_duart->b_tx_cb().set(m_modem, FUNC(rs232_port_device::write_txd));
+	m_modem->rxd_handler().set(m_duart, FUNC(mc68681_device::rx_b_w));
+	// DUART IP1 = CTS
+	m_modem->cts_handler().set(m_duart, FUNC(mc68681_device::ip1_w));
+	// DUART OP1 = RTS
+	m_duart->outport_cb().append(m_modem, FUNC(rs232_port_device::write_rts)).bit(1);
```

Comment:
We generally prefer to have devices’ output routing grouped with their instantiation.  Please move the `m_duart` configuration up to group it with the instantiation, before instantiating the RS232 ports.  You can refer to `m_terminal` and `m_modem` before the devices are instantiated.

Also, do you need to configure the default serial settings for the terminal or other devices?  If so, you can do it with `DEVICE_INPUT_DEFAULTS_START`, `set_option_device_input_defaults`, etc. (see src/mame/intel/intellec4.cpp for an example).

Reply:
Thanks! The terminal defaults (9600 8N1) work just fine. Should I still specify them?

Reply:
If 9600 8N1 works you don’t need to override it.  The defaults aren’t going to change.

Reply:
About that, I think it would be useful to have something like set_option_device_input_defaults("*", ...)
In case of rs232, if the driver expects serial comms to be at 1200 baud for example, it applies to all of the slot options.